### PR TITLE
Allow users to specify a custom output length when resampling

### LINKED ISSUE #13: Allow specifying expected output length when resampling
Hi Alexandre :)

Consider this scenario:

We have a machine learning model that takes in and outputs audio at a high sample rate. For some reasons (amongst other things execution time) the model uses an internal sample rate that is lower than the input-output. The output shape is expected to be the same as the input shape.

```
class ResampleWrapper(nn.Module):
    """
    This class downsamples audio before it's passed to the audio denoiser model,
    and upsamples the audio back to the original sample rate before returning.
    """

    def __init__(
        self,
        model: nn.Module,
        input_sample_rate: int = 48_000,
        internal_sample_rate: int = 32_000,
    ):
        super().__init__()
        self.model = model
        self.input_sample_rate = input_sample_rate
        self.internal_sample_rate = internal_sample_rate

        self.downsampler = ResampleFrac(
            self.input_sample_rate, self.internal_sample_rate
        )
        self.upsampler = ResampleFrac(self.internal_sample_rate, self.input_sample_rate)

    def forward(self, x):
        """
        :param x: tensor with shape (batch_size, num_channels, num_samples)
        :return: tensor with shape (batch_size, num_channels, num_samples)
        """
        x = self.downsampler(x)
        x = self.model(x)
        x = self.upsampler(x)
        return x
```

Depending on the exact length of the input, downsampling and then upsampling will often give an output that is off by one sample in length.

In `librosa` this kind of issue can be solved by setting the [fix parameter](https://librosa.org/doc/main/generated/librosa.resample.html) or by using [fix_length](https://librosa.org/doc/main/generated/librosa.util.fix_length.html#librosa.util.fix_length) manually.

I imagine that `julius.ResampleFrac` could provide a parameter called something like `expected_length` in its forward function that customizes the slice end offset in the end so that the result has the given length 😄

Would you like a pull request that adds this feature?

### PR DESCRIPTION
**Pull request by iver56**
Close #13

### CODE REVIEW

File: julius/resample.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 import torch
 from torch.nn import functional as F
+from typing import Optional
```
Comment (adefossez) [inline julius/resample.py:19]: nit: imports from standard library goes together in the paragraph above.

File: tests/test_lowpass.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def delta(a, b, ref, fraction=0.9):
     return 100 * th.abs(a - b).mean() / ref.std()
 
 
-TOLERANCE = 1  # Tolerence to errors as percentage of the std of the input signal
+TOLERANCE = 1  # Tolerance to errors as percentage of the std of the input signal
```
Comment (adefossez) [inline tests/test_lowpass.py:22]: thanks for the fix

**Review by adefossez — CHANGES_REQUESTED** [review]
Thanks a lot for the PR. After thinking more on this I think things are more complex than just choosing between `floor` and `ceil`. For instance, let say I do the following

```python
x = torch.randn(19)
y = julius.resample_frac(x, 7, 1)  # y.shape is [2]
x2 = julius.resample_frac(y, 1, 7). # x2 shape is [14]
```
Now I want to have a way to make `x2` of the same size as `x`. With the current diff, this is not possible. I can pick `y` to be of length `3`, but I cannot take x2 of length 19 (only 21, or 22).

In fact, given an input of length L, we know that the true input length (given as a float, for instance if it is itself the output of a previous resampliong) is between `(L-1, L]` (for instance the ideal length of y would be 2.71), so that the true output length as a float could be anywhere between `((L-1) * new / old, L * new / old]`, and the possible actual output lengths could be anything in `(floor((L-1) * new) / old, ceil(L * new / old)]` (i would say the left side of the segment is excluded, because it would actually be the same as the right side for an input that is smaller).

nit: if you have time, I would also add a `full` boolean parameter that automatically pick the longest possible output without having the user doing the computation. After doing arbitrary sequences of resampling, it is always possible to just keep the first L steps, rather than computing all the intermediate sizes that will make it work. If you do that, we could actually chose to be more permissive in the previous definition of the available segment, for instance to do something like:

```python
x = torch.randn(19)
y = julius.resample_frac(x, 7, 1, full=True) #  length is 3
z = julius.resample_frac(x, 5, 1, full=True) # length is 1
y2 = julius.resample_frac(x, 1, 5)  # length is 5
x2 = julius.resample_frac(x, 1, 7, length=len(x))  # normally 19 shouldn't be acceptable here
```

In that case the segment would be [0, ceil(L * new / old)], which would be convenient but maybe a bit too permissive. What do you think?

**Comment by iver56** [comment]
Good point. I'll make the range check more permissive. But should output_length=0 really be allowed? That seems counter-intuitive.

I prefer to leave the `full` parameter for future work

**Comment by adefossez** [comment]
thanks for the update!

**Comment by adefossez** [comment]
I'll push a new version of julius to PyPI tomorrow.

**Comment by iver56** [comment]
Sweet, thanks for the swift response!

**Comment by adefossez** [comment]
I made some minor edits https://github.com/adefossez/julius/commit/3804ada9ecd61d1566dfd786714cfc2ec55eb3f4 and I added the `full` parameter.
The new version has been released to PyPI as 0.2.5.

**Pull request merged**